### PR TITLE
#379 total before tax logic

### LINKED ISSUE #379: Total before tax logic

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #379 

and some refactoring....... 👀

### CODE REVIEW

File: server/service/src/invoice_line/inbound_shipment_line/update/generate.rs [hunk 1]
```diff
@@ -94,13 +95,14 @@ fn generate_line(
...
-            update_line.cost_price_per_pack * update_line.number_of_packs as f64;
-    }
-
-    if let Some(tax) = tax {
-        update_line.tax = tax.percentage;
+    } else if let Some(number_of_packs) = number_of_packs {
+        update_line.total_before_tax = update_line.cost_price_per_pack * number_of_packs as f64;
+    } else if let Some(cost_price_per_pack) = cost_price_per_pack {
+        update_line.total_before_tax = cost_price_per_pack * update_line.number_of_packs as f64;
+    } else if let (Some(number_of_packs), Some(cost_price_per_pack)) =
```
Comment (mark-prins) [inline server/service/src/invoice_line/inbound_shipment_line/update/generate.rs:102]: this block will have to be first - if either `number_of_packs` or `cost_price_per_pack` aren't None then you won't reach the final if block - or is there some magic that I'm missing here?

Comment (roxy-dao) [reply server/service/src/invoice_line/inbound_shipment_line/update/generate.rs:98]: [0d8193d](https://github.com/openmsupply/open-msupply/pull/380/commits/0d8193d5e69eeaa1f10113ac6d3e607be3e60e7e)

**Review by mark-prins — APPROVED** [review]
👍 LGTM

File: server/service/src/invoice_line/inbound_shipment_line/update/generate.rs [hunk 2]
```diff
@@ -94,13 +95,14 @@ fn generate_line(
 
     if let Some(total_before_tax) = total_before_tax {
```
Comment (andreievg) [inline server/service/src/invoice_line/inbound_shipment_line/update/generate.rs:96]: I would simplify this slightly with:
```rust
  update_line.total_before_tax = if let Some(total_before_tax) = total_before_tax {
        total_before_tax
    } else if number_of_packs.is_some() || cost_price_per_pack.is_some() {
        update_line.cost_price_per_pack * update_line.number_of_packs as f64
    } else {
        update_line.total_after_tax
    };
```
`update_line.total_before_tax =` -> indicates the result of conditions to always be new total_before_tax

`  update_line.cost_price_per_pack * update_line.number_of_packs` -> number of packs and cost_price_per_pack already set to new values.

**Review by andreievg — COMMENTED** [review]
Some suggestions, and also the tests don't compile

File: server/service/src/invoice_line/inbound_shipment_line/update/generate.rs (hunk 2 above)
Comment (roxy-dao) [reply server/service/src/invoice_line/inbound_shipment_line/update/generate.rs:96]: Would the last else statement not be pdate_line.cost_price_per_pack * update_line.number_of_packs as f64? 🤔

**Pull request merged**